apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex

```diff
@@ -0,0 +1,377 @@
+           gas: non_neg_integer(),
+           gas_price: non_neg_integer(),
+           call_data: binary(),
+           call_stack: [non_neg_integer()]
+         }
+
+   @type t :: %ContractCallTx{
+           contract: Identifier.t(),
+           vm_version: integer(),
+           amount: non_neg_integer(),
+           gas: non_neg_integer(),
+           gas_price: non_neg_integer(),
+           call_data: binary(),
+           call_stack: [non_neg_integer()]
+         }
+
+   @type tx_type_state() :: Chainstate.calls()
+
+   defstruct [
+     :contract,
+     :vm_version,
+     :amount,
+     :gas,
+     :gas_price,
+     :call_data,
+     :call_stack
+   ]
+
+   use ExConstructor
+
+   defguardp is_non_neg_integer(value) when is_integer(value) and value >= 0
+
+   @spec get_chain_state_name() :: :calls
+   def get_chain_state_name, do: :calls
+
+   @spec init(payload()) :: t()
+   def init(%{
+         contract: %Identifier{} = identified_contract,
+         vm_version: vm_version,
+         amount: amount,
+         gas: gas,
+         gas_price: gas_price,
+         call_data: call_data,
+         call_stack: call_stack
+       }) do
+     %ContractCallTx{
+       contract: identified_contract,
+       vm_version: vm_version,
+       amount: amount,
+       gas: gas,
+       gas_price: gas_price,
+       call_data: call_data,
+       call_stack: call_stack
+     }
+   end
+
+   def init(%{
+         contract: contract,
+         vm_version: vm_version,
+         amount: amount,
+         gas: gas,
+         gas_price: gas_price,
+         call_data: call_data,
+         call_stack: call_stack
+       }) do
+     identified_contract = Identifier.create_identity(contract, :contract)
+
+     %ContractCallTx{
+       contract: identified_contract,
+       vm_version: vm_version,
+       amount: amount,
+       gas: gas,
+       gas_price: gas_price,
+       call_data: call_data,
+       call_stack: call_stack
+     }
+   end
+
+   @spec validate(t(), DataTx.t()) :: :ok | {:error, String.t()}
+   def validate(
+         %ContractCallTx{
+           contract: contract
+         },
+         _data_tx
+       ) do
+     if validate_identifier(contract, :contract) do
+       :ok
+     else
+       {:error, "#{__MODULE__}: Invalid contract address: #{inspect(contract)}"}
+     end
+   end
+
+   @spec process_chainstate(
+           Chainstate.accounts(),
+           tx_type_state(),
+           non_neg_integer(),
+           t(),
+           DataTx.t(),
+           Transaction.context()
+         ) :: {:ok, {Chainstate.accounts(), tx_type_state()}}
+   def process_chainstate(
+         accounts,
+         calls,
+         block_height,
+         %ContractCallTx{} = call_tx,
+         data_tx,
+         context
+       ) do
+     # Transfer the attached funds to the callee, before the calling of the contract
+     sender = DataTx.main_sender(data_tx)
+     nonce = DataTx.nonce(data_tx)
+
+     chain_state = Chain.chain_state()
```

Comment:
You can't call `Chain` from tx. `Chain.chain_state` gives you chainstate of top block. Block we are validating currently doesn't have to have top_block as `prev`. (We might be calculating chainstate of competing fork)

Reply:
I see that you need this to pass to `run_contract`. You should change the `process_chainstate` signature in some way instead. I'm not sure what's the best solution. We decided previously to pass `accounts` and some `tx_type_state` instead of everything to simplify `process_chainstate` functions, but passing those two is not enough for this tx.

Reply:
This applies also to other txs.